File: backend/routers/handover.py
import logging
from typing import Literal, Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from backend.auth import get_current_user
from backend.db import get_db_connection
from backend.services import handover as handover_service

logger = logging.getLogger("kabadilink.routers.handover")
router = APIRouter(tags=["Handover"])


class OtpVerifyRequest(BaseModel):
    code: str
    actual_weight_kg: Optional[float] = None


class StageOfflineRequest(BaseModel):
    actual_weight_kg: float = Field(..., gt=0)


class StatusUpdateRequest(BaseModel):
    status: Literal["EN_ROUTE", "COMPLETED"]


class PaymentRequest(BaseModel):
    amount: float = Field(..., gt=0)
    method: Literal["CASH", "DIGITAL"]


@router.get("/handover/{lot_id}")
def get_handover(lot_id: str, current_user: dict = Depends(get_current_user)):
    with get_db_connection() as conn:
        return handover_service.get_handover(conn, lot_id)


@router.post("/handover/{lot_id}/otp/generate")
def generate_otp(lot_id: str, current_user: dict = Depends(get_current_user)):
    with get_db_connection() as conn:
        return handover_service.generate_otp(conn, lot_id, current_user["id"])


@router.post("/handover/{lot_id}/otp/verify")
def verify_otp(lot_id: str, payload: OtpVerifyRequest, current_user: dict = Depends(get_current_user)):
    with get_db_connection() as conn:
        return handover_service.verify_otp(conn, lot_id, payload.code, current_user["id"], payload.actual_weight_kg)


@router.post("/handover/{lot_id}/stage-offline")
def stage_offline(lot_id: str, payload: StageOfflineRequest, current_user: dict = Depends(get_current_user)):
    with get_db_connection() as conn:
        return handover_service.stage_offline(conn, lot_id, current_user["id"], payload.actual_weight_kg)


@router.post("/handover/{lot_id}/status")
def update_status(lot_id: str, payload: StatusUpdateRequest, current_user: dict = Depends(get_current_user)):
    with get_db_connection() as conn:
        return handover_service.update_status(conn, lot_id, payload.status, current_user["id"])


@router.post("/handover/{lot_id}/payment")
def record_payment(lot_id: str, payload: PaymentRequest, current_user: dict = Depends(get_current_user)):
    with get_db_connection() as conn:
        return handover_service.record_payment(conn, lot_id, current_user["id"], payload.amount, payload.method)


@router.get("/passport/{lot_id}")
def get_passport(lot_id: str, current_user: dict = Depends(get_current_user)):
    with get_db_connection() as conn:
        return handover_service.get_passport(conn, lot_id)

File: frontend/kabadiwala-connect-main/src/lib/api/handover.ts
import { request } from "./client";
import { Handover, HandoverStatus, Passport, Transaction, EprRecord } from "../../types/api";

export async function getHandover(lotId: string): Promise<Handover> {
  return request<Handover>(`/handover/${lotId}`, {
    method: "GET",
  });
}

export async function generateHandoverOtp(lotId: string): Promise<{ otp_code: string }> {
  return request<{ otp_code: string }>(`/handover/${lotId}/otp/generate`, {
    method: "POST",
  });
}

export async function verifyHandoverOtp(
  lotId: string,
  code: string,
  actualWeightKg?: number | null
): Promise<Handover> {
  return request<Handover>(`/handover/${lotId}/otp/verify`, {
    method: "POST",
    body: {
      code,
      actual_weight_kg: actualWeightKg,
    },
  });
}

export async function stageOfflineHandover(
  lotId: string,
  actualWeightKg: number
): Promise<Handover> {
  return request<Handover>(`/handover/${lotId}/stage-offline`, {
    method: "POST",
    body: {
      actual_weight_kg: actualWeightKg,
    },
  });
}

export async function updateHandoverStatus(
  lotId: string,
  status: "EN_ROUTE" | "COMPLETED"
): Promise<Handover> {
  return request<Handover>(`/handover/${lotId}/status`, {
    method: "POST",
    body: { status },
  });
}

export async function recordHandoverPayment(
  lotId: string,
  amount: number,
  method: "CASH" | "DIGITAL"
): Promise<Transaction> {
  return request<Transaction>(`/handover/${lotId}/payment`, {
    method: "POST",
    body: {
      amount,
      method,
    },
  });
}

export async function getPassport(lotId: string): Promise<Passport> {
  return request<Passport>(`/passport/${lotId}`, {
    method: "GET",
  });
}

export async function getEprRecord(lotId: string): Promise<EprRecord> {
  return request<EprRecord>(`/lots/${lotId}/epr-record`, {
    method: "GET",
  });
}
